File: java/Print_in_Order_1114.java


import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

class Foo_1 {

  private AtomicInteger firstJobDone = new AtomicInteger(0);
  private AtomicInteger secondJobDone = new AtomicInteger(0);

  public Foo_1() {}

  public void first(Runnable printFirst) throws InterruptedException {
    // printFirst.run() outputs "first".
    printFirst.run();
    // mark the first job as done, by increasing its count.
    firstJobDone.incrementAndGet();
  }

  public void second(Runnable printSecond) throws InterruptedException {
    while (firstJobDone.get() != 1) {
      // waiting for the first job to be done.
    }
    // printSecond.run() outputs "second".
    printSecond.run();
    // mark the second as done, by increasing its count.
    secondJobDone.incrementAndGet();
  }

  public void third(Runnable printThird) throws InterruptedException {
    while (secondJobDone.get() != 1) {
      // waiting for the second job to be done.
    }
    // printThird.run() outputs "third".
    printThird.run();
  }
}

class Foo_2 {

  Semaphore run2, run3;

  public Foo_2() {
    run2 = new Semaphore(0);
    run3 = new Semaphore(0);
  }

  Runnable printFirst =
      new Runnable() {
        public void run() {
          System.out.print("first");
          run2.release();
        }
      };

  Runnable printSecond =
      new Runnable() {
        public void run() {
          try {
            run2.acquire();
          } catch (InterruptedException e) {
            System.out.println("Error");
          }
          System.out.print("second");
          run3.release();
        }
      };

  Runnable printThird =
      new Runnable() {
        public void run() {
          try {
            run3.acquire();
          } catch (InterruptedException e) {
            System.out.println("Error");
          }
          System.out.print("third");
        }
      };
  

  public static void main(String[] args) {
    Foo_2 bar = new Foo_2();

    Thread one = new Thread(bar.printFirst);
    Thread two = new Thread(bar.printSecond);
    Thread three = new Thread(bar.printThird);

    one.start();
    two.start();
    three.start();
  }
  
}

File: python/Print_in_Order_1114.py
# -*- coding: utf-8 -*-
"""
"""

from threading import Lock
from threading import Lock
from threading import Semaphore


class Foo_1:
    def __init__(self):
        self.firstJobDone = Lock()
        self.secondJobDone = Lock()
        self.firstJobDone.acquire()
        self.secondJobDone.acquire()

    def first(self, printFirst: 'Callable[[], None]') -> None:
        # printFirst() outputs "first".
        printFirst()
        # Notify the thread that is waiting for the first job to be done.
        self.firstJobDone.release()

    def second(self, printSecond: 'Callable[[], None]') -> None:
        # Wait for the first job to be done
        with self.firstJobDone:
            # printSecond() outputs "second".
            printSecond()
            # Notify the thread that is waiting for the second job to be done.
            self.secondJobDone.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        # Wait for the second job to be done.
        with self.secondJobDone:
            # printThird() outputs "third".
            printThird()


"""
Start with two closed gates represented by 0-value semaphores. 
Second and third thread are waiting behind these gates.
When the first thread prints, it opens the gate for the second thread.
When the second thread prints, it opens the gate for the third thread.
"""


class Foo_2:
    def __init__(self):
        self.gates = (Semaphore(0), Semaphore(0))

    def first(self, printFirst):
        printFirst()
        self.gates[0].release()

    def second(self, printSecond):
        with self.gates[0]:
            printSecond()
            self.gates[1].release()

    def third(self, printThird):
        with self.gates[1]:
            printThird()


"""
Start with two locked locks. First thread unlocks the first lock that the second thread is waiting on. 
Second thread unlocks the second lock that the third thread is waiting on.
"""


class Foo_3:
    def __init__(self):
        self.locks = (Lock(), Lock())
        self.locks[0].acquire()
        self.locks[1].acquire()

    def first(self, printFirst):
        printFirst()
        self.locks[0].release()

    def second(self, printSecond):
        with self.locks[0]:
            printSecond()
            self.locks[1].release()

    def third(self, printThird):
        with self.locks[1]:
            printThird()
